File: python/vcp_gateway/taint.py
"""Taint labels and propagation (SPEC §12).

Every datum entering or leaving the Planner carries exactly one label. Derived
data inherits the MOST restrictive label of its sources. Authority MUST NOT flow
from ``untrusted_*`` data, and classified data movement to external sinks can be
forbidden even when the model proposes it.

This module reproduces ``conformance/vectors/taint.json`` exactly:
propagation (most-restrictive), ``AUTHORITY_FROM_TAINTED_DATA``, and
``DATA_FLOW_FORBIDDEN``.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Iterable, Optional

__all__ = [
    "RESTRICTIVENESS_ORDER",
    "Label",
    "most_restrictive",
    "authority_decision",
    "data_flow_decision",
    "Decision",
]

# Most restrictive first (index 0) to least restrictive last, per the vector's
# `restrictiveness_order_most_to_least`.
RESTRICTIVENESS_ORDER = (
    "secret",
    "untrusted_tool_result",
    "untrusted_resource_data",
    "policy_only",
    "trusted_manifest_summary",
    "user_instruction",
    "developer_instruction",
    "system_instruction",
)

# Labels from which authority MUST NOT flow (§12).
_NON_AUTHORITATIVE = {"untrusted_resource_data", "untrusted_tool_result"}

# Classifications that may not move to an external sink.
_RESTRICTED_CLASSIFICATIONS = {"confidential", "secret", "restricted"}

Label = str


@dataclass(frozen=True)
class Decision:
    decision: str  # "allow" | "deny"
    reason_code: Optional[str] = None

    def as_dict(self) -> dict:
        d: dict = {"decision": self.decision}
        if self.reason_code is not None:
            d["reason_code"] = self.reason_code
        return d


def _rank(label: Label) -> int:
    try:
        return RESTRICTIVENESS_ORDER.index(label)
    except ValueError as exc:  # pragma: no cover - defensive
        raise ValueError(f"unknown taint label: {label!r}") from exc


def most_restrictive(labels: Iterable[Label]) -> Label:
    """Return the most restrictive label among ``labels`` (§12 lattice)."""
    labels = list(labels)
    if not labels:
        raise ValueError("at least one source label is required")
    # Smallest index == most restrictive.
    return min(labels, key=_rank)


def authority_decision(label: Label, authorizes: bool) -> Decision:
    """Decide whether a datum with ``label`` may authorize an action (§12).

    ``authorizes=True`` means the datum is being used to justify/authorize an
    action. Authority from an ``untrusted_*`` label MUST be denied with
    ``AUTHORITY_FROM_TAINTED_DATA``. Using such data purely AS DATA
    (``authorizes=False``) is allowed.
    """
    if authorizes and label in _NON_AUTHORITATIVE:
        return Decision("deny", "AUTHORITY_FROM_TAINTED_DATA")
    return Decision("allow")


def data_flow_decision(
    *,
    classification: Optional[str],
    sink: str,
    from_: Optional[str] = None,
    to: Optional[str] = None,
) -> Decision:
    """Decide whether a classified datum may flow to ``sink`` (§12, §16).

    Classified data moving to an ``external`` sink is forbidden
    (``DATA_FLOW_FORBIDDEN``). Movement to an ``internal-metadata`` sink (e.g.
    event title/time/attendees on a calendar) is permitted — the §16 worked
    example: email→calendar metadata is allowed, email→slack external is not.
    """
    if classification in _RESTRICTED_CLASSIFICATIONS and sink == "external":
        return Decision("deny", "DATA_FLOW_FORBIDDEN")
    return Decision("allow")

File: rust/crates/vcp-gateway/src/taint.rs
//! Taint / data-flow engine (§12).
//!
//! Reproduces `conformance/vectors/taint.json`:
//!
//! - **Most-restrictive propagation:** derived data inherits the most
//!   restrictive label among its sources.
//! - **Authority rule:** authority derived from an `untrusted_*` label is
//!   denied (`AUTHORITY_FROM_TAINTED_DATA`).
//! - **Data-flow rule:** moving classified data to an external sink is forbidden
//!   (`DATA_FLOW_FORBIDDEN`); the same data may flow into internal capability
//!   *metadata*.

/// The eight taint labels (§12), ordered most-restrictive (index 0) to
/// least-restrictive. This is the lattice ordering the spec and `taint.json`
/// agree on (`restrictiveness_order_most_to_least`).
pub const RESTRICTIVENESS_MOST_TO_LEAST: [&str; 8] = [
    "secret",
    "untrusted_tool_result",
    "untrusted_resource_data",
    "policy_only",
    "trusted_manifest_summary",
    "user_instruction",
    "developer_instruction",
    "system_instruction",
];

/// Rank of a label: 0 = most restrictive. Unknown labels are treated as maximally
/// restrictive (fail closed).
fn rank(label: &str) -> usize {
    RESTRICTIVENESS_MOST_TO_LEAST
        .iter()
        .position(|l| *l == label)
        .unwrap_or(0)
}

/// True if a label denotes untrusted, non-authoritative data (§12).
pub fn is_untrusted(label: &str) -> bool {
    label == "untrusted_resource_data" || label == "untrusted_tool_result"
}

/// Propagate taint: a derived datum inherits the MOST restrictive source label
/// (§12). Returns `None` for an empty source set.
pub fn propagate<'a>(sources: &[&'a str]) -> Option<&'a str> {
    sources.iter().copied().min_by_key(|l| rank(l))
}

/// Decision for an authority check or data-flow check.
#[derive(Debug, Clone, PartialEq, Eq)]
pub enum TaintDecision {
    Allow,
    Deny(&'static str),
}

/// Authority rule (§12): a datum with the given label is being used to authorize
/// an action iff `authorizes` is true. Authority from any `untrusted_*` label
/// MUST be denied with `AUTHORITY_FROM_TAINTED_DATA`. Using untrusted data merely
/// as *data* (not authority) is allowed.
pub fn check_authority(label: &str, authorizes: bool) -> TaintDecision {
    if authorizes && is_untrusted(label) {
        TaintDecision::Deny("AUTHORITY_FROM_TAINTED_DATA")
    } else {
        TaintDecision::Allow
    }
}

/// A declared data movement to evaluate (mirrors a `taint.json` dataflow case).
pub struct DataFlow<'a> {
    pub from: &'a str,
    pub to: &'a str,
    pub classification: &'a str,
    /// `"external"` or `"internal-metadata"`.
    pub sink: &'a str,
}

/// Data-flow rule (§12): classified data to an external sink is forbidden
/// (`DATA_FLOW_FORBIDDEN`); the same classified data flowing into internal
/// capability metadata (title/time/attendees) is allowed.
pub fn check_data_flow(flow: &DataFlow) -> TaintDecision {
    let classified = matches!(flow.classification, "confidential" | "personal");
    if classified && flow.sink == "external" {
        TaintDecision::Deny("DATA_FLOW_FORBIDDEN")
    } else {
        TaintDecision::Allow
    }
}
